File: swim-system-java/swim-core-java/swim.dataflow/src/main/java/swim/dataflow/AbstractRecordOutlet.java
// Copyright 2015-2021 Swim inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.dataflow;

import java.util.Iterator;
import swim.collections.HashTrieMap;
import swim.streamlet.Inlet;
import swim.streamlet.KeyEffect;
import swim.streamlet.KeyOutlet;
import swim.streamlet.MapInlet;
import swim.streamlet.Outlet;
import swim.streamlet.StreamletContext;
import swim.streamlet.StreamletScope;
import swim.structure.Field;
import swim.structure.Item;
import swim.structure.Record;
import swim.structure.Text;
import swim.structure.Value;
import swim.util.Cursor;

public abstract class AbstractRecordOutlet extends Record implements RecordOutlet {

  protected HashTrieMap<Value, KeyEffect> effects;
  protected HashTrieMap<Value, KeyOutlet<Value, Value>> outlets;
  protected Inlet<? super Record>[] outputs;
  protected int version;

  public AbstractRecordOutlet() {
    this.effects = HashTrieMap.empty();
    this.outlets = HashTrieMap.empty();
    this.outputs = null;
    this.version = -1;
  }

  @Override
  public StreamletScope<? extends Value> streamletScope() {
    return null;
  }

  @Override
  public StreamletContext streamletContext() {
    final StreamletScope<? extends Value> scope = this.streamletScope();
    if (scope != null) {
      return scope.streamletContext();
    }
    return null;
  }

  public boolean containsOwnKey(Value key) {
    return this.containsKey(key);
  }

  @Override
  public abstract Iterator<Value> keyIterator();

  @Override
  public Record get() {
    return this;
  }

  @Override
  public Outlet<Value> outlet(Value key) {
    if (!this.containsOwnKey(key)) {
      final StreamletScope<? extends Value> scope = this.streamletScope();
      if (scope instanceof RecordOutlet && ((RecordOutlet) scope).containsKey(key)) {
        // TODO: Support dynamic shadowing?
        return ((RecordOutlet) scope).outlet(key);
      }
    }
    KeyOutlet<Value, Value> outlet = this.outlets.get(key);
    if (outlet == null) {
      outlet = new KeyOutlet<Value, Value>(this, key);
      this.outlets = this.outlets.updated(key, outlet);
      this.decohereInputKey(key, KeyEffect.UPDATE);
    }
    return outlet;
  }

  @Override
  public Outlet<Value> outlet(String key) {
    return this.outlet(Text.from(key));
  }

  @Override
  public Iterator<Inlet<? super Record>> outputIterator() {
    return this.outputs != null ? Cursor.array(this.outputs) : Cursor.empty();
  }

  @SuppressWarnings("unchecked")
  @Override
  public void bindOutput(Inlet<? super Record> output) {
    final Inlet<? super Record>[] oldOutputs = this.outputs;
    final int n = oldOutputs != null ? oldOutputs.length : 0;
    final Inlet<? super Record>[] newOutputs = (Inlet<? super Record>[]) new Inlet<?>[n + 1];
    if (n > 0) {
      System.arraycopy(oldOutputs, 0, newOutputs, 0, n);
    }
    newOutputs[n] = output;
    this.outputs = newOutputs;
  }

  @SuppressWarnings("unchecked")
  @Override
  public void unbindOutput(Inlet<? super Record> output) {
    final Inlet<? super Record>[] oldOutputs = this.outputs;
    final int n = oldOutputs != null ? oldOutputs.length : 0;
    for (int i = 0; i < n; i += 1) {
      if (oldOutputs[i] == output) {
        if (n > 1) {
          final Inlet<? super Record>[] newOutputs = (Inlet<? super Record>[]) new Inlet<?>[n - 1];
          System.arraycopy(oldOutputs, 0, newOutputs, 0, i);
          System.arraycopy(oldOutputs, i + 1, newOutputs, i, (n - 1) - i);
          this.outputs = newOutputs;
        } else {
          this.outputs = null;
        }
        break;
      }
    }
  }

  @Override
  public void unbindOutputs() {
    final HashTrieMap<Value, KeyOutlet<Value, Value>> outlets = this.outlets;
    if (!outlets.isEmpty()) {
      this.outlets = HashTrieMap.empty();
      final Iterator<KeyOutlet<Value, Value>> keyOutlets = outlets.valueIterator();
      while (keyOutlets.hasNext()) {
        final KeyOutlet<Value, Value> keyOutlet = keyOutlets.next();
        keyOutlet.unbindOutputs();
      }
    }
    final Inlet<? super Record>[] oldOutputs = this.outputs;
    if (oldOutputs != null) {
      this.outputs = null;
      for (int i = 0, n = oldOutputs.length; i < n; i += 1) {
        oldOutputs[i].unbindInput();
      }
    }
  }

  @Override
  public void disconnectOutputs() {
    final HashTrieMap<Value, KeyOutlet<Value, Value>> outlets = this.outlets;
    if (!outlets.isEmpty()) {
      this.outlets = HashTrieMap.empty();
      final Iterator<KeyOutlet<Value, Value>> keyOutlets = outlets.valueIterator();
      while (keyOutlets.hasNext()) {
        final KeyOutlet<Value, Value> keyOutlet = keyOutlets.next();
        keyOutlet.disconnectOutputs();
      }
    }
    final Inlet<? super Record>[] outputs = this.outputs;
    if (outputs != null) {
      this.outputs = null;
      for (int i = 0, n = outputs.length; i < n; i += 1) {
        final Inlet<? super Record> output = outputs[i];
        output.unbindInput();
        output.disconnectOutputs();
      }
    }
    for (Item member : this) {
      if (member instanceof Field) {
        member = member.toValue();
      }
      if (member instanceof RecordOutlet) {
        ((RecordOutlet) member).disconnectOutputs();
      } else if (member instanceof RecordStreamlet) {
        ((RecordStreamlet) member).disconnectOutputs();
      }
    }
  }

  @Override
  public void disconnectInputs() {
    // nop
  }

  @SuppressWarnings("unchecked")
  @Override
  public void decohereInputKey(Value key, KeyEffect effect) {
    final HashTrieMap<Value, KeyEffect> oldEffects = this.effects;
    if (oldEffects.get(key) != effect) {
      this.willDecohereInputKey(key, effect);
      this.effects = oldEffects.updated(key, effect);
      this.version = -1;
      this.onDecohereInputKey(key, effect);
      final int n = this.outputs != null ? this.outputs.length : 0;
      for (int i = 0; i < n; i += 1) {
        final Inlet<?> output = this.outputs[i];
        if (output instanceof MapInlet<?, ?, ?>) {
          ((MapInlet<Value, Value, ? super Record>) output).decohereOutputKey(key, effect);
        } else {
          output.decohereOutput();
        }
      }
      final KeyOutlet<Value, Value> outlet = this.outlets.get(key);
      if (outlet != null) {
        outlet.decohereInput();
      }
      this.didDecohereInputKey(key, effect);
    }
  }

  @Override
  public void decohereInput() {
    if (this.version >= 0) {
      this.willDecohereInput();
      this.version = -1;
      this.onDecohereInput();
      final int n = this.outputs != null ? this.outputs.length : 0;
      for (int i = 0; i < n; i += 1) {
        this.outputs[i].decohereOutput();
      }
      final Iterator<KeyOutlet<Value, Value>> outlets = this.outlets.valueIterator();
      while (outlets.hasNext()) {
        outlets.next().decohereInput();
      }
      this.didDecohereInput();
    }
  }

  @SuppressWarnings("unchecked")
  @Override
  public void recohereInputKey(Value key, int version) {
    if (this.version < 0) {
      final HashTrieMap<Value, KeyEffect> oldEffects = this.effects;
      final KeyEffect effect = oldEffects.get(key);
      if (effect != null) {
        this.willRecohereInputKey(key, effect, version);
        this.effects = oldEffects.removed(key);
        this.onRecohereInputKey(key, effect, version);
        for (int i = 0, n = this.outputs != null ? this.outputs.length : 0; i < n; i += 1) {
          final Inlet<?> output = this.outputs[i];
          if (output instanceof MapInlet<?, ?, ?>) {
            ((MapInlet<Value, Value, ? super Record>) output).recohereOutputKey(key, version);
          }
        }
        final KeyOutlet<Value, Value> outlet = this.outlets.get(key);
        if (outlet != null) {
          outlet.recohereInput(version);
        }
        this.didRecohereInputKey(key, effect, version);
      }
    }
  }

  @Override
  public void recohereInput(int version) {
    if (this.version < 0) {
      this.willRecohereInput(version);
      final Iterator<Value> keys = this.effects.keyIterator();
      while (keys.hasNext()) {
        this.recohereInputKey(keys.next(), version);
      }
      this.version = version;
      this.onRecohereInput(version);
      for (int i = 0, n = this.outputs != null ? this.outputs.length : 0; i < n; i += 1) {
        this.outputs[i].recohereOutput(version);
      }
      for (Item member : this) {
        if (member instanceof Field) {
          member = member.toValue();
        }
        if (member instanceof RecordOutlet) {
          ((RecordOutlet) member).recohereInput(version);
        } else if (member instanceof RecordStreamlet) {
          ((RecordStreamlet) member).recohere(version);
        }
      }
      this.didRecohereInput(version);
    }
  }

  protected void willDecohereInputKey(Value key, KeyEffect effect) {
    // hook
  }

  protected void onDecohereInputKey(Value key, KeyEffect effect) {
    // hook
  }

  protected void didDecohereInputKey(Value key, KeyEffect effect) {
    // hook
  }

  protected void willDecohereInput() {
    // hook
  }

  protected void onDecohereInput() {
    // hook
  }

  protected void didDecohereInput() {
    // hook
  }

  protected void willRecohereInputKey(Value key, KeyEffect effect, int version) {
    // hook
  }

  protected void onRecohereInputKey(Value key, KeyEffect effect, int version) {
    // hook
  }

  protected void didRecohereInputKey(Value key, KeyEffect effect, int version) {
    // hook
  }

  protected void willRecohereInput(int version) {
    // hook
  }

  protected void onRecohereInput(int version) {
    // hook
  }

  protected void didRecohereInput(int version) {
    // hook
  }

}

File: swim-js/swim-runtime/swim-core/swim-dataflow/src/main/AbstractRecordOutlet.ts
// Copyright 2015-2022 Swim.inc
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Mutable, Arrays, Cursor, Map} from "@swim/util";
import {BTree} from "@swim/collections";
import {AnyItem, Item, Field, AnyValue, Value, Record, Text, Selector} from "@swim/structure";
import {
  Inlet,
  Outlet,
  KeyEffect,
  MapInlet,
  MapOutlet,
  MapOutletCombinators,
  KeyOutlet,
  StreamletContext,
  StreamletScope,
  FilterFieldsFunction,
  FilterFieldsCombinator,
} from "@swim/streamlet";
import {RecordOutlet} from "./RecordOutlet";
import {RecordStreamlet} from "./"; // forward import

/** @public */
export abstract class AbstractRecordOutlet extends Record implements RecordOutlet {
  constructor() {
    super();
    this.effects = new BTree();
    this.outlets = new BTree();
    this.outputs = Arrays.empty;
    this.version = -1;
  }

  /** @internal */
  readonly effects: BTree<Value, KeyEffect>;

  /** @internal */
  readonly outlets: BTree<Value, KeyOutlet<Value, Value>>;

  /** @internal */
  readonly outputs: ReadonlyArray<Inlet<Record>>;

  /** @internal */
  readonly version: number;

  declare readonly streamletScope: StreamletScope<Value> | null; // getter defined below to work around useDefineForClassFields lunacy

  get streamletContext(): StreamletContext | null {
    const scope = this.streamletScope;
    if (scope !== null) {
      return scope.streamletContext;
    }
    return null;
  }

  hasOwn(key: AnyValue): boolean {
    return this.has(key);
  }

  override get(): Record;
  override get(key: AnyValue): Value;
  override get(key?: AnyValue): Record | Value {
    if (key === void 0) {
      return this;
    } else {
      return super.get(key);
    }
  }

  abstract keyIterator(): Cursor<Value>;

  outlet(key: Value | string): Outlet<Value> {
    if (typeof key === "string") {
      key = Text.from(key);
    }
    if (!this.hasOwn(key)) {
      const scope = this.streamletScope;
      if (RecordOutlet.is(scope) && scope.has(key)) {
        // TODO: Support dynamic shadowing?
        return scope.outlet(key);
      }
    }
    const oldOutlets = this.outlets;
    let outlet = oldOutlets.get(key);
    if (outlet === void 0) {
      outlet = new KeyOutlet<Value, Value>(this, key);
      (this as Mutable<this>).outlets = oldOutlets.updated(key, outlet);
      this.decohereInputKey(key, KeyEffect.Update);
    }
    return outlet;
  }

  outputIterator(): Cursor<Inlet<Record>> {
    return Cursor.array(this.outputs);
  }

  bindOutput(output: Inlet<Record>): void {
    (this as Mutable<this>).outputs = Arrays.inserted(output, this.outputs);
  }

  unbindOutput(output: Inlet<Record>): void {
    (this as Mutable<this>).outputs = Arrays.removed(output, this.outputs);
  }

  unbindOutputs(): void {
    const oldOutlets = this.outlets;
    if (oldOutlets.isEmpty()) {
      (this as Mutable<this>).outlets = new BTree();
      oldOutlets.forEach(function (key: Value, keyOutlet: KeyOutlet<Value, Value>) {
        keyOutlet.unbindOutputs();
      }, this);
    }
    const oldOutputs = this.outputs;
    (this as Mutable<this>).outputs = Arrays.empty;
    for (let i = 0, n = oldOutputs.length; i < n; i += 1) {
      const output = oldOutputs[i]!;
      output.unbindInput();
    }
  }

  disconnectOutputs(): void {
    const oldOutlets = this.outlets;
    if (oldOutlets.isEmpty()) {
      (this as Mutable<this>).outlets = new BTree();
      oldOutlets.forEach(function (key: Value, keyOutlet: KeyOutlet<Value, Value>) {
        keyOutlet.disconnectOutputs();
      }, this);
    }
    const oldOutputs = this.outputs;
    (this as Mutable<this>).outputs = Arrays.empty;
    for (let i = 0, n = oldOutputs.length; i < n; i += 1) {
      const output = oldOutputs[i]!;
      output.unbindInput();
      output.disconnectOutputs();
    }
    this.forEach(function (member: Item): void {
      if (member instanceof Field) {
        member = member.toValue();
      }
      if (member instanceof AbstractRecordOutlet) {
        member.disconnectOutputs();
      } else if (member instanceof RecordStreamlet) {
        member.disconnectOutputs();
      } else if (RecordOutlet.is(member)) {
        member.disconnectOutputs();
      }
    }, this);
  }

  disconnectInputs(): void {
    // nop
  }

  decohereInputKey(key: Value, effect: KeyEffect): void {
    const oldEffects = this.effects;
    if (oldEffects.get(key) !== effect) {
      this.willDecohereInputKey(key, effect);
      (this as Mutable<this>).effects = oldEffects.updated(key, effect);
      (this as Mutable<this>).version = -1;
      this.onDecohereInputKey(key, effect);
      const outputs = this.outputs;
      for (let i = 0, n = outputs.length; i < n; i += 1) {
        const output = outputs[i]!;
        if (MapInlet.is(output)) {
          output.decohereOutputKey(key, effect);
        } else {
          output.decohereOutput();
        }
      }
      const outlet = this.outlets.get(key);
      if (outlet !== void 0) {
        outlet.decohereInput();
      }
      this.didDecohereInputKey(key, effect);
    }
  }

  decohereInput(): void {
    if (this.version >= 0) {
      this.willDecohereInput();
      (this as Mutable<this>).version = -1;
      this.onDecohereInput();
      const outputs = this.outputs;
      for (let i = 0, n = outputs.length; i < n; i += 1) {
        const output = outputs[i]!;
        output.decohereOutput();
      }
      this.outlets.forEach(function (key: Value, outlet: KeyOutlet<Value, Value>): void {
        outlet.decohereInput();
      }, this);
      this.didDecohereInput();
    }
  }

  recohereInputKey(key: Value, version: number): void {
    if (this.version < 0) {
      const oldEffects = this.effects;
      const effect = oldEffects.get(key);
      if (effect !== void 0) {
        this.willRecohereInputKey(key, effect, version);
        (this as Mutable<this>).effects = oldEffects.removed(key);
        this.onRecohereInputKey(key, effect, version);
        const outputs = this.outputs;
        for (let i = 0, n = outputs.length; i < n; i += 1) {
          const output = outputs[i];
          if (MapInlet.is(output)) {
            output.recohereOutputKey(key, version);
          }
        }
        const outlet = this.outlets.get(key);
        if (outlet !== void 0) {
          outlet.recohereInput(version);
        }
        this.didRecohereInputKey(key, effect, version);
      }
    }
  }

  recohereInput(version: number): void {
    if (this.version < 0) {
      this.willRecohereInput(version);
      this.effects.forEach(function (key: Value): void {
        this.recohereInputKey(key, version);
      }, this);
      (this as Mutable<this>).version = version;
      this.onRecohereInput(version);
      const outputs = this.outputs;
      for (let i = 0, n = outputs.length; i < n; i += 1) {
        const output = outputs[i]!;
        output.recohereOutput(version);
      }
      this.forEach(function (member: Item): void {
        if (member instanceof Field) {
          member = member.toValue();
        }
        if (member instanceof AbstractRecordOutlet) {
          member.recohereInput(version);
        } else if (member instanceof RecordStreamlet) {
          member.recohere(version);
        } else if (RecordOutlet.is(member)) {
          member.recohereInput(version);
        }
      }, this);
      this.didRecohereInput(version);
    }
  }

  protected willDecohereInputKey(key: Value, effect: KeyEffect): void {
    // hook
  }

  protected onDecohereInputKey(key: Value, effect: KeyEffect): void {
    // hook
  }

  protected didDecohereInputKey(key: Value, effect: KeyEffect): void {
    // hook
  }

  protected willDecohereInput(): void {
    // hook
  }

  protected onDecohereInput(): void {
    // hook
  }

  protected didDecohereInput(): void {
    // hook
  }

  protected willRecohereInputKey(key: Value, effect: KeyEffect, version: number): void {
    // hook
  }

  protected onRecohereInputKey(key: Value, effect: KeyEffect, version: number): void {
    // hook
  }

  protected didRecohereInputKey(key: Value, effect: KeyEffect, version: number): void {
    // hook
  }

  protected willRecohereInput(version: number): void {
    // hook
  }

  protected onRecohereInput(version: number): void {
    // hook
  }

  protected didRecohereInput(version: number): void {
    // hook
  }

  override filter(predicate?: AnyItem): Selector;
  override filter(func: FilterFieldsFunction<Value, Value>): MapOutlet<Value, Value, Map<Value, Value>>;
  override filter(func: AnyItem | FilterFieldsFunction<Value, Value>): Selector | MapOutlet<Value, Value, Map<Value, Value>> {
    if (typeof func !== "function") {
      return super.filter(func as AnyItem);
    } else {
      const combinator = new FilterFieldsCombinator<Value, Value, Record>(func);
      combinator.bindInput(this);
      return combinator;
    }
  }
}
Object.defineProperty(AbstractRecordOutlet.prototype, "streamletScope", {
  get(this: AbstractRecordOutlet): StreamletScope<Value> | null {
    return null;
  },
  configurable: true,
});
/** @public */
export interface AbstractRecordOutlet extends MapOutletCombinators<Value, Value, Record> {
}
MapOutletCombinators.define(AbstractRecordOutlet.prototype);
